library/src/options.cxx

```diff
@@ -10,197 +12,201 @@
     }
-    catch (const std::out_of_range&)
+    catch (std::out_of_range const& ex)
     {
-      log::error("Option ", name, " does not exist");
+      throw options::parsing_exception("Cannot parse " + str + " into a ratio as it would go out of double range");
     }
   }
 
-  template<typename T>
-  void get(const std::string& name, T& value) const
+  void parse(std::string str, std::string& value)
+  {
+    value = str;
+  }
+
+  void parse(std::string str, std::vector<double>& value)
   {
+    // TODO implement proper parsing
     try
     {
-      value = std::get<T>(this->Options.at(name));
+      std::istringstream split(str);
+      value.clear();
+      for (std::string each; std::getline(split, each, ','); value.push_back(std::stod(each)));
     }
-    catch (const std::bad_variant_access&)
+    catch (std::invalid_argument const& ex)
     {
-      log::error("Trying to get option ", name, " with incompatible type");
-      return;
+      throw options::parsing_exception("Cannot parse " + str + " into a double vector");
     }
-    catch (const std::out_of_range&)
+    catch (std::out_of_range const& ex)
     {
-      log::error("Option ", name, " does not exist");
-      return;
+      throw options::parsing_exception("Cannot parse " + str + " into a double vector as it would go out of double range");
     }
   }
 
-  template<typename T>
-  T get(const std::string& name) const
+  void setAsString(const std::string& name, std::string str)
   {
-    T val = {};
-    this->get(name, val);
-    return val;
+    option_variant_t var = options_struct_internals::get(this->OptionsStruct, name);
+
+    // TODO Use std::visit ?
+    if (std::holds_alternative<bool>(var))
+    {
+      bool value;
+      this->parse(str, value);
+      var = value;
+    }
+    else if (std::holds_alternative<int>(var))
+    {
+      int value;
+      this->parse(str, value);
+      var = value;
+    }
+    else if (std::holds_alternative<double>(var))
+    {
+      double value;
+      this->parse(str, value);
+      var = value;
+    }
+    else if (std::holds_alternative<ratio_t>(var))
+    {
+      ratio_t value(0);
+      this->parse(str, value);
+      var = value;
+    }
+    else if (std::holds_alternative<std::string>(var))
+    {
+      std::string value;
+      this->parse(str, value);
+      var = value;
+    }
+    else if (std::holds_alternative<std::vector<double>>(var))
+    {
+      std::vector<double> value;
+      this->parse(str, value);
+      var = value;
+    }
+    options_struct_internals::set(this->OptionsStruct, name, var);
   }
 
-  template<typename T>
-  T& getRef(const std::string& name)
+  std::string getAsString(const std::string& name)
   {
+    option_variant_t var = options_struct_internals::get(this->OptionsStruct, name);
+    std::string str;
     try
     {
-      return std::get<T>(this->Options.at(name));
+      if (std::holds_alternative<bool>(var))
```

Comment:
what about using `std::visit` instead? All these `if/else` looks odd to me.

Not tested but it could be something like:

```cpp
// generic impl
template<typename T>
std::string toString(const T& v)
{
    return std::to_string(v);
}

// specialized impl for strings
std::string toString(const std::string& v)
{
    return v;
}

// specialized impl for vectors
std::string toString([const std::vector<double>& v)
{
   std::string r;
    unsigned int i = 0;
    for (auto& elem : vec)
    {
      r+= ((i > 0) ? "," : "");
      r+= toString(elem);
      i++;
    }
   return ret;
}

//

std::string getAsString(const std::string& name)
{
    option_variant_t var = options_struct_internals::get(this->OptionsStruct, name);
    return std::visit([](const auto& e) { return toString(e); }, var);
}
```

`toString` can be `class TypeStringConverter` and implement an `operator()` if you prefer callable objects.